### DISCUSSION TITLE
Text classification with Mutually Exclusive classes

### DISCUSSION THREAD

**Discussion by koaning** [explosion/spaCy#6613] [Help: Coding & Implementations] [answered]
I'm trying to run a benchmark on a classification problem using spacy-nightly rc2. The dataset has classes that represent intents in a chatbot setting and these intents can be assumed to be mutually exclusive. 

I took code from [here](https://github.com/explosion/projects/tree/v3/tutorials/textcat_docs_issues) to serve as an inspiration and I ended up with a config for the classification like below. 

```cfg
[components.textcat.model]
@architectures = "spacy.TextCatEnsemble.v1"
exclusive_classes = true
pretrained_vectors = null
width = 64
conv_depth = 2
embed_size = 2000
window_size = 1
ngram_size = 1
dropout = null
nO = null
```

I'm showing this part of the config file to show that I've explicitly set `exclusive_classes = true`. I've also made a small script to get the data in the right format. 

```python
import typer
import srsly
from pathlib import Path
from spacy.tokens import DocBin
import spacy

CATEGORIES = ['pin_change', 'restaurant_reservation', 'current_location', 'uber', ... 'freeze_account']

def main(
    input_path: Path = typer.Argument(..., exists=True, dir_okay=False),
    output_path: Path = typer.Argument(..., dir_okay=False),
):
    nlp = spacy.blank("en")
    doc_bin = DocBin()
    data_tuples = ((eg["text"], eg) for eg in srsly.read_jsonl(input_path))
    for doc, eg in nlp.pipe(data_tuples, as_tuples=True):
        doc.cats = {category: 0 for category in CATEGORIES}
        doc.cats[eg["label"]] = 1
        doc_bin.add(doc)
    doc_bin.to_disk(output_path)
    print(f"Processed {len(doc_bin)} documents: {output_path.name}")


if __name__ == "__main__":
    typer.run(main)
````

When I run this pipeline, it seems to be doing something wrong. 

```
E    #       LOSS TOK2VEC  LOSS TEXTCAT  CATS_SCORE  SCORE 
---  ------  ------------  ------------  ----------  ------
  0       0          0.00          0.01        0.00    0.00
  0     200          0.00          1.35        0.00    0.00
  0     400          0.00          0.90        0.00    0.00
  0     600          0.00          0.59        0.00    0.00
  0     800          0.00          0.39        0.00    0.00
  1    1000          0.00          0.26        0.00    0.00
  1    1200          0.00          0.17        0.00    0.00
  1    1400          0.00          0.11        0.00    0.00
  2    1600          0.00          0.08        0.00    0.00
  3    1800          0.00          0.05        0.00    0.00
  3    2000          0.00          0.03        0.00    0.00
```

So I'm wondering if perhaps I'm missing something.

**Comment by koaning** [comment, marked as answer]
I think I've solved this by not using the config from the project but instead generating one on the docs page [here](https://spacy.io/usage/training#quickstart) and using the command line app to finish it. 

```
python -m spacy init fill-config
```

It now seems that everything is running better.

```
E    #       LOSS TOK2VEC  LOSS TEXTCAT  CATS_SCORE  SCORE 
---  ------  ------------  ------------  ----------  ------
  0       0          0.00          0.01        0.00    0.00
  0     200          0.00          1.33        0.00    0.00
  0     400          0.00          0.83        1.93    0.02
  0     600          0.00          0.47       13.32    0.13
  0     800          0.00          0.26       29.98    0.30
  1    1000          0.00          0.14       44.43    0.44
  1    1200          0.00          0.07       54.62    0.55
  1    1400          0.00          0.04       62.81    0.63
```

**Reply by svlandeg** [reply]
Hey Vincent. Just as a bit of background, the pipeline component that used to be called `textcat` previously did multi labeling by default. Now it's the other way around and `textcat` does single-label by default. If you want a multi-label, you'd need to use the pipeline component `textcat_multilabel`.

We've done this pretty much for the reason highlighted by your posts: if by accident (or some bug somewhere) you run the multi-label one on single-label data, it'll end up performing poorly but the user might not necessarily notice what's up. With the default flipped, you'll get a custom error warning telling you to use `textcat_multilabel` if you try to run the `textcat` with multiple labels. So the error won't go unnoticed and hopefully everyone will have better textcat results overall ;-) 

I'm still not sure what went wrong in your original experiment as it did look like you set the boolean correctly. But happy to hear it's fixed nonetheless ;-)

**Reply by koaning** [reply]
Actually! I think I changed one extra thing. 

```python
def main(
    input_path: Path = typer.Argument(..., exists=True, dir_okay=False),
    output_path: Path = typer.Argument(..., dir_okay=False),
):
    nlp = spacy.blank("en")
    doc_bin = DocBin()
    data_tuples = ((eg["text"], eg) for eg in srsly.read_jsonl(input_path))
    for doc, eg in nlp.pipe(data_tuples, as_tuples=True):
        # I turned off this bit. 
        # doc.cats = {category: 0 for category in CATEGORIES}
        doc.cats[eg["label"]] = 1
        doc_bin.add(doc)
    doc_bin.to_disk(output_path)
    print(f"Processed {len(doc_bin)} documents: {output_path.name}")
```

**Reply by koaning** [reply]
For those interested, I've placed the project [here](https://github.com/koaning/spacy-intent-example).
